raft/src/main/java/org/apache/kafka/raft/KafkaRaftClient.java

```diff
@@ -1428,51 +1433,66 @@ private Records readCommitted(OffsetAndEpoch offsetAndEpoch, long highWatermark)
     }
 
     @Override
-    public void shutdown(int timeoutMs) {
-        // TODO: Safe to access epoch? Need to reset connections to be able to send EndQuorumEpoch? Block until shutdown completes?
+    public CompletableFuture<Void> shutdown(int timeoutMs) {
+        logger.info("Beginning graceful shutdown");
+        CompletableFuture<Void> shutdownComplete = new CompletableFuture<>();
         kafkaRaftMetrics.close();
-        shutdown.set(new GracefulShutdown(timeoutMs, quorum.epoch()));
+        shutdown.set(new GracefulShutdown(timeoutMs, quorum.epoch(), shutdownComplete));
         channel.wakeup();
+        return shutdownComplete;
     }
 
     public OptionalLong highWatermark() {
         return quorum.highWatermark();
     }
 
     private class GracefulShutdown {
+        final long shutdownTimeoutMs;
         final int epoch;
         final Timer finishTimer;
-
-        final AtomicBoolean finished = new AtomicBoolean(false);
+        final CompletableFuture<Void> completeFuture;
 
         public GracefulShutdown(long shutdownTimeoutMs,
-                                int epoch) {
+                                int epoch,
+                                CompletableFuture<Void> completeFuture) {
+            this.shutdownTimeoutMs = shutdownTimeoutMs;
```

Comment:
nit: we keep this variable local only for logging purposes. Maybe we can just keep it in `Timer` so that it can return `timeoutMs` instead?

Reply:
We don't currently keep a reference to the timeout inside `Timer`, but perhaps we could. Let me consider this.